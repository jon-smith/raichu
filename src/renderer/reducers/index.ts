import { combineReducers } from 'redux';
import { useSelector } from 'react-redux';

import { ActivityState, activityReducer } from './activityReducer';

export interface RootState {
	activities: ActivityState;
}

export const rootReducer = combineReducers<RootState | undefined>({
	activities: activityReducer
});

export const useRootSelector = <T extends {}>(selector: (s: RootState) => T) =>
	useSelector<RootState, T>(s => selector(s));

export const useActivitySelector = <T extends {}>(selector: (s: ActivityState) => T) =>
	useSelector<RootState, T>(s => selector(s.activities));
