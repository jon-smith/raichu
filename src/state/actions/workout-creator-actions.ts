import { Action } from 'redux';

export const SET_INTERVALS = 'SET_WORKOUT_CREATOR_INTERVALS';
export const SET_SELECTED = 'SET_WORKOUT_CREATOR_SELECTED_INTERVAL';
export const UNDO = 'WORKOUT_CREATOR_UNDO';
export const REDO = 'WORKOUT_CREATOR_REDO';

export type Interval = {
	intensity: number;
	length: number;
	color: string;
};

export interface SetIntervalsAction extends Action {
	type: 'SET_WORKOUT_CREATOR_INTERVALS';
	intervals: Interval[];
}

export interface SetSelectedIntervalAction extends Action {
	type: 'SET_WORKOUT_CREATOR_SELECTED_INTERVAL';
	index: number | null;
}

export interface UndoAction extends Action {
	type: 'WORKOUT_CREATOR_UNDO';
}

export interface RedoAction extends Action {
	type: 'WORKOUT_CREATOR_REDO';
}

export const setIntervals = (intervals: Interval[]): SetIntervalsAction => ({
	type: SET_INTERVALS,
	intervals
});

export const setSelectedIndex = (index: number | null): SetSelectedIntervalAction => ({
	type: SET_SELECTED,
	index
});

export const undo = (): UndoAction => ({
	type: UNDO
});

export const redo = (): RedoAction => ({
	type: REDO
});

export type WorkoutCreatorAction =
	| SetIntervalsAction
	| SetSelectedIntervalAction
	| UndoAction
	| RedoAction;
