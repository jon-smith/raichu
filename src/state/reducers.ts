import { combineReducers } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { ActivityState, reducer as activityReducer } from './activity-data/activity-data-slice';
import { ViewState, reducer as viewReducer } from './view/view-slice';
import {
	WorkoutCreatorState,
	reducer as workoutCreatorReducer
} from './workout-creator/workout-creator-slice';

export const rootReducer = combineReducers({
	activities: activityReducer,
	view: viewReducer,
	workoutCreator: workoutCreatorReducer
});

export type RootState = ReturnType<typeof rootReducer>;

export const useRootSelector = <T extends {}>(selector: (s: RootState) => T) =>
	useSelector<RootState, T>(s => selector(s));

export const useActivitySelector = <T extends {}>(selector: (s: ActivityState) => T) =>
	useSelector<RootState, T>(s => selector(s.activities));

export const useViewSelector = <T extends {}>(selector: (s: ViewState) => T) =>
	useSelector<RootState, T>(s => selector(s.view));

export const useWorkoutCreatorSelector = <T extends {}>(selector: (s: WorkoutCreatorState) => T) =>
	useSelector<RootState, T>(s => selector(s.workoutCreator));
