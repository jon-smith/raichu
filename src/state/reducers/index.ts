import { combineReducers } from 'redux';
import { useSelector } from 'react-redux';

import { ActivityState, activityReducer } from './activityReducer';
import { ViewState, viewReducer } from './viewReducer';

export interface RootState {
	activities: ActivityState;
	view: ViewState;
}

export const rootReducer = combineReducers<RootState | undefined>({
	activities: activityReducer,
	view: viewReducer
});

export const useRootSelector = <T extends {}>(selector: (s: RootState) => T) =>
	useSelector<RootState, T>(s => selector(s));

export const useActivitySelector = <T extends {}>(selector: (s: ActivityState) => T) =>
	useSelector<RootState, T>(s => selector(s.activities));

export const useViewSelector = <T extends {}>(selector: (s: ViewState) => T) =>
	useSelector<RootState, T>(s => selector(s.view));
