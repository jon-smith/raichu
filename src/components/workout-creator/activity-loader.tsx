import * as React from 'react';
import { useCallback, useMemo, useState } from 'react';
import * as d3 from 'd3';

import Box, { BoxProps } from '@material-ui/core/Box';
import IconButton from '@material-ui/core/IconButton';
import { Close } from '@material-ui/icons';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import FormGroup from '@material-ui/core/FormGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Switch from '@material-ui/core/Switch';

import XYPlot, { DataSeriesT } from 'generic-components/charts/xy-plot';
import ActivityFileDrop, {
	FileAndData,
	extractActivityData,
} from 'generic-components/file/activity-file-drop';

import { formatSecondsAsHHMMSS } from 'library/utils/time-format-utils';
import { buildNiceTimeTicksToDisplay } from 'library/utils/chart-utils';
import { getAttributes } from 'library/activity-data/activity-container';

import { useDispatchCallback } from 'store/dispatch-hooks';
import { setActivity, clearActivity } from 'store/workout-creator/slice';
import { getDetectedIntervals } from 'store/workout-creator/selectors';
import { useWorkoutCreatorSelector } from 'store/reducers';

import SettingsFormGroup from './settings-form-group';

const ActivityLoadHeader = () => {
	const { loadedActivity } = useWorkoutCreatorSelector((s) => ({
		loadedActivity: s.activity,
		isGenerating: s.generatingFromActivity,
		ftp: s.ftp,
	}));

	const setActivityDispatcher = useDispatchCallback(setActivity);
	const clearActivitityDispatcher = useDispatchCallback(clearActivity);

	const addFiles = useCallback(
		(files: FileAndData[]) => {
			if (files.length > 0) {
				const { data } = files[0];
				const activities = extractActivityData(data);
				if (activities.length > 0) {
					setActivityDispatcher(activities[0]);
				}
			}
		},
		[setActivityDispatcher]
	);

	if (loadedActivity) {
		return (
			<>
				<IconButton aria-label="close file" onClick={() => clearActivitityDispatcher()}>
					<Close />
				</IconButton>
				<div style={{ marginRight: '20px' }}>
					<span>{`Loaded activity: ${getAttributes(loadedActivity).name}`}</span>
				</div>
			</>
		);
	}

	return <ActivityFileDrop onAddFiles={addFiles} allowMultiple={false} />;
};

const stopClickFocusPropagation: Partial<BoxProps> = {
	onClick: (event) => event.stopPropagation(),
	onFocus: (event) => event.stopPropagation(),
};

function buildPowerSeries(data: DataSeriesT['data']): DataSeriesT {
	return {
		name: 'power-vs-time',
		data,
		seriesType: 'line',
		color: '#551A8B',
	};
}

function buildDiscrepencyCurveSeries(data: DataSeriesT['data']): DataSeriesT {
	return {
		name: 'discrepency-curve',
		data,
		seriesType: 'line',
		color: '#ff6961',
	};
}

function buildDetectedStepsSeries(data: DataSeriesT['data']): DataSeriesT {
	return {
		name: 'detected-steps',
		data,
		seriesType: 'mark',
		color: 'red',
	};
}

const FormSwitch = (params: {
	label: string;
	value: boolean;
	setValue: (v: boolean) => void;
	name: string;
	color?: 'primary' | 'secondary';
}) => {
	const { label, value, setValue, name, color } = params;
	return (
		<FormControlLabel
			control={
				<Switch
					checked={value}
					onChange={(e) => setValue(e.target.checked)}
					name={name}
					color={color}
				/>
			}
			label={label}
		/>
	);
};

const ActivityChart = () => {
	const [showDiscrepencyCurve, setShowDiscrepencyCurve] = useState(false);
	const [showDetectedSteps, setShowDetectedSteps] = useState(false);
	const [showSmoothedInput, setShowSmoothedInput] = useState(false);

	const {
		rawInput,
		smoothedInput,
		discrepencyCurve,
		stepThreshold,
		detectedStepTimePoints,
	} = useWorkoutCreatorSelector((s) => ({
		...getDetectedIntervals(s),
		stepThreshold: s.generationParams.stepThreshold,
	}));

	const powerData = showSmoothedInput ? smoothedInput : rawInput;

	const maxPower = useMemo(() => d3.max(rawInput, (p) => p.y ?? 0) ?? 0, [rawInput]);

	const powerDataSeries = useMemo(() => buildPowerSeries(powerData), [powerData]);

	const discrepencyCurveSeries = useMemo(() => {
		const absolute = discrepencyCurve.map((d) => ({ x: d.t, y: Math.abs(d.delta) }));
		const maxDelta = d3.max(absolute, (a) => a.y) ?? 0;
		const adjustment = maxDelta > 0 ? maxPower / maxDelta : 1.0;
		const normalised = absolute.map((a) => ({
			...a,
			y: a.y > stepThreshold ? a.y * adjustment : null,
		}));
		return buildDiscrepencyCurveSeries(normalised);
	}, [discrepencyCurve, maxPower, stepThreshold]);

	const detectedStepsSeries = useMemo(() => {
		const stepVsPower = detectedStepTimePoints.map((d) => ({ x: d, y: powerData[d].y }));
		return buildDetectedStepsSeries(stepVsPower);
	}, [detectedStepTimePoints, powerData]);

	const activityXTicks = useMemo(() => {
		const maxX = powerData.length > 0 ? powerData[powerData.length - 1].x : 3600;
		return buildNiceTimeTicksToDisplay(maxX, 5);
	}, [powerData]);

	const allSeries = useMemo(
		() => [
			...(showDiscrepencyCurve ? [discrepencyCurveSeries] : []),
			powerDataSeries,
			...(showDetectedSteps ? [detectedStepsSeries] : []),
		],
		[
			detectedStepsSeries,
			discrepencyCurveSeries,
			powerDataSeries,
			showDetectedSteps,
			showDiscrepencyCurve,
		]
	);

	return (
		<Box display="flex" flexDirection="column" width="100%">
			<FormGroup row>
				<FormSwitch
					label="Show Smoothed Power"
					name="showSmooth"
					color="secondary"
					value={showSmoothedInput}
					setValue={setShowSmoothedInput}
				/>
				<FormSwitch
					label="Show Discrepency Curve"
					name="showDCurve"
					color="secondary"
					value={showDiscrepencyCurve}
					setValue={setShowDiscrepencyCurve}
				/>
				<FormSwitch
					label="Show Steps"
					name="showSteps"
					value={showDetectedSteps}
					setValue={setShowDetectedSteps}
				/>
			</FormGroup>
			<Box className="workout-creator-activity-data-chart">
				<XYPlot
					series={allSeries}
					xTickFormat={formatSecondsAsHHMMSS}
					xTickValues={activityXTicks}
					xAxisLabel="time"
					yAxisLabel="Power"
				/>
			</Box>
		</Box>
	);
};

const ActivityLoader = () => (
	<>
		<ExpansionPanel>
			<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
				<Box display="flex" flexDirection="column" {...stopClickFocusPropagation}>
					<Box
						display="flex"
						flexDirection="row"
						alignItems="center"
						minHeight="2em"
						width="100%"
						{...stopClickFocusPropagation}
					>
						<ActivityLoadHeader />
					</Box>
					<Box width="100%" {...stopClickFocusPropagation}>
						<SettingsFormGroup />
					</Box>
				</Box>
			</ExpansionPanelSummary>
			<ExpansionPanelDetails>
				<ActivityChart />
			</ExpansionPanelDetails>
		</ExpansionPanel>
	</>
);

export default ActivityLoader;
