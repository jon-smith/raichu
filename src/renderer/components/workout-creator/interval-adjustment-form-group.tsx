import * as React from 'react';
import { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { batchActions } from 'redux-batched-actions';

import { remote } from 'electron';
import * as fs from 'fs';

import Box from '@material-ui/core/Box';
import FormGroup from '@material-ui/core/FormGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import TextField from '@material-ui/core/TextField';
import IconButton from '@material-ui/core/IconButton';
import { Undo, Redo, Add, Save, Delete } from '@material-ui/icons';
import { createStyles, makeStyles } from '@material-ui/core/styles';

import * as moment from 'moment';
import TimePicker from 'rc-time-picker';

import { useWorkoutCreatorSelector } from 'state/reducers';
import { useDispatchCallback } from 'state/actions';
import {
	Interval,
	canUndo,
	canRedo,
	selectedOrNewInterval,
	intervalsWithColor,
	actions as WorkoutCreatorActions
} from 'state/workout-creator/workout-creator-slice';
import { buildMRCFileString } from 'shared/activity-data/export-mrc';

const useStyles = makeStyles(theme =>
	createStyles({
		root: {
			padding: theme.spacing(0, 2),
			width: 160
		},
		label: {
			padding: theme.spacing(1)
		}
	})
);

const useActions = () => {
	const setIntervals = useDispatchCallback(WorkoutCreatorActions.setIntervals);
	const undo = useDispatchCallback(WorkoutCreatorActions.undo);
	const redo = useDispatchCallback(WorkoutCreatorActions.redo);
	const setSelectedIntensity = useDispatchCallback(WorkoutCreatorActions.setSelectedIntensity);
	const setSelectedLength = useDispatchCallback(WorkoutCreatorActions.setSelectedLength);

	return {
		setIntervals,
		undo,
		redo,
		setSelectedIntensity,
		setSelectedLength
	};
};

const loadFileDialog = (intervals: readonly Interval[]) => {
	const savePath = remote.dialog.showSaveDialogSync({
		title: 'Save workout as MRC',
		filters: [{ name: 'MRC', extensions: ['mrc'] }]
	});

	if (savePath) {
		fs.writeFileSync(
			savePath,
			buildMRCFileString(
				'Workout',
				'',
				intervals.map(i => ({ durationSeconds: i.length, intensityPercent: i.intensity * 100 }))
			)
		);
	}
};

const IntervalAdjustmentFormGroup = () => {
	const {
		intervals,
		selectedIndex,
		currentSelectedInterval,
		undoEnabled,
		redoEnabled,
		newInterval
	} = useWorkoutCreatorSelector(w => ({
		intervals: intervalsWithColor(w),
		selectedIndex: w.selectedIndex,
		undoEnabled: canUndo(w),
		redoEnabled: canRedo(w),
		newInterval: w.newInterval,
		currentSelectedInterval: selectedOrNewInterval(w)
	}));

	const { setIntervals, undo, redo, setSelectedIntensity, setSelectedLength } = useActions();

	const saveToMRC = useCallback(() => loadFileDialog(intervals), [intervals]);

	const setIntervalDuration = useCallback(
		(m: moment.Moment | null) => {
			const seconds = m?.unix() ?? 0;
			setSelectedLength(seconds);
		},
		[setSelectedLength]
	);

	const addInterval = useCallback(() => {
		if (newInterval.length > 0) {
			setIntervals([...intervals, newInterval]);
		}
	}, [newInterval, setIntervals, intervals]);

	const deleteSelected = useCallback(() => {
		if (selectedIndex != null) {
			const copy = intervals.slice();
			copy.splice(selectedIndex, 1);
			setIntervals(copy);
		}
	}, [intervals, selectedIndex, setIntervals]);

	const formControlClasses = useStyles();

	return (
		<FormGroup row>
			<IconButton aria-label="save to MRC" onClick={saveToMRC}>
				<Save />
			</IconButton>
			<IconButton aria-label="undo" disabled={!undoEnabled} onClick={undo}>
				<Undo />
			</IconButton>
			<IconButton aria-label="redo" disabled={!redoEnabled} onClick={redo}>
				<Redo />
			</IconButton>
			<FormControlLabel
				classes={formControlClasses}
				control={
					<TimePicker
						value={moment.utc(currentSelectedInterval.length * 1000)}
						onChange={d => setIntervalDuration(d)}
						showHour={false}
						showMinute={true}
						showSecond={true}
					/>
				}
				label="Duration"
				labelPlacement="start"
			/>
			<FormControlLabel
				classes={formControlClasses}
				control={
					<TextField
						type="number"
						variant="outlined"
						size="small"
						margin="none"
						value={currentSelectedInterval.intensity * 100}
						onChange={e => setSelectedIntensity(parseFloat(e.target.value) * 0.01)}
					/>
				}
				label="Intensity"
				labelPlacement="start"
			/>
			<IconButton
				aria-label="add"
				onClick={addInterval}
				disabled={newInterval.length === 0 || selectedIndex !== null}
			>
				<Add />
			</IconButton>
			<IconButton aria-label="delete" onClick={deleteSelected} disabled={selectedIndex === null}>
				<Delete />
			</IconButton>
		</FormGroup>
	);
};

export default IntervalAdjustmentFormGroup;
