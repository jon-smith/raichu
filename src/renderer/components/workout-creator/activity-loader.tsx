import * as React from 'react';
import { useCallback, useMemo, useState } from 'react';
import * as d3 from 'd3';

import Box, { BoxProps } from '@material-ui/core/Box';
import IconButton from '@material-ui/core/IconButton';
import { Close } from '@material-ui/icons';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import FormGroup from '@material-ui/core/FormGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Switch from '@material-ui/core/Switch';

import XYPlot, { DataSeriesT } from 'ui/charts/xy-plot';
import GpxFileDrop, { FileAndGpx } from 'ui/file/gpx-file-drop';

import { formatSecondsAsHHMMSS } from 'shared/utils/time-format-utils';
import { buildNiceTimeTicksToDisplay } from 'shared/utils/chart-utils';
import { getAttributes, fromGPXData } from 'shared/activity-data/activity-container';

import { useDispatchCallback } from 'state/dispatch-hooks';
import { setActivity, clearActivity } from 'state/workout-creator/slice';
import {
	getActivityProcessedPowerTimeSeries,
	getMovingWindowDiscrepencyCurve,
	getDetectedStepTimePoints
} from 'state/workout-creator/selectors';
import { useWorkoutCreatorSelector } from 'state/reducers';

import SettingsFormGroup from './settings-form-group';

const ActivityLoadHeader = () => {
	const { loadedActivity } = useWorkoutCreatorSelector(s => ({
		loadedActivity: s.activity,
		isGenerating: s.generatingFromActivity,
		ftp: s.ftp
	}));

	const setActivityDispatcher = useDispatchCallback(setActivity);
	const clearActivitityDispatcher = useDispatchCallback(clearActivity);

	const addFiles = useCallback(
		(files: FileAndGpx[]) => {
			if (files.length > 0) {
				const { gpx } = files[0];
				setActivityDispatcher(fromGPXData(gpx));
			}
		},
		[setActivityDispatcher]
	);

	if (loadedActivity) {
		return (
			<>
				<IconButton aria-label="close file" onClick={() => clearActivitityDispatcher()}>
					<Close />
				</IconButton>
				<div style={{ marginRight: '20px' }}>
					<span>{`Loaded activity: ${getAttributes(loadedActivity).name}`}</span>
				</div>
			</>
		);
	}

	return <GpxFileDrop onAddFiles={addFiles} allowMultiple={false} />;
};

const stopClickFocusPropagation: Partial<BoxProps> = {
	onClick: event => event.stopPropagation(),
	onFocus: event => event.stopPropagation()
};

function buildPowerSeries(data: DataSeriesT['data']): DataSeriesT {
	return {
		name: 'power-vs-time',
		data,
		seriesType: 'line',
		color: '#551A8B'
	};
}

function buildDiscrepencyCurveSeries(data: DataSeriesT['data']): DataSeriesT {
	return {
		name: 'discrepency-curve',
		data,
		seriesType: 'line',
		color: '#ff6961'
	};
}

function buildDetectedStepsSeries(data: DataSeriesT['data']): DataSeriesT {
	return {
		name: 'detected-steps',
		data,
		seriesType: 'mark',
		color: 'red'
	};
}

const ActivityLoader = () => {
	const [showDiscrepencyCurve, setShowDiscrepencyCurve] = useState(false);
	const [showDetectedSteps, setShowDetectedSteps] = useState(false);

	const { powerData, discrepencyCurve, stepThreshold, detectedSteps } = useWorkoutCreatorSelector(
		s => ({
			powerData: getActivityProcessedPowerTimeSeries(s),
			discrepencyCurve: getMovingWindowDiscrepencyCurve(s),
			detectedSteps: getDetectedStepTimePoints(s),
			stepThreshold: s.generationParams.stepThreshold
		})
	);

	const powerDataSeries = useMemo(() => buildPowerSeries(powerData), [powerData]);

	const maxPower = useMemo(() => d3.max(powerData, p => p.y ?? 0) ?? 0, [powerData]);

	const discrepencyCurveSeries = useMemo(() => {
		const absolute = discrepencyCurve.map(d => ({ x: d.t, y: Math.abs(d.delta) }));
		const maxDelta = d3.max(absolute, a => a.y) ?? 0;
		const adjustment = maxDelta > 0 ? maxPower / maxDelta : 1.0;
		const normalised = absolute.map(a => ({
			...a,
			y: a.y > stepThreshold ? a.y * adjustment : null
		}));
		return buildDiscrepencyCurveSeries(normalised);
	}, [discrepencyCurve, maxPower, stepThreshold]);

	const detectedStepsSeries = useMemo(() => {
		const stepVsPower = detectedSteps.map(d => ({ x: d, y: powerData[d].y }));
		return buildDetectedStepsSeries(stepVsPower);
	}, [detectedSteps, powerData]);

	const activityXTicks = useMemo(() => {
		const maxX = powerData.length > 0 ? powerData[powerData.length - 1].x : 3600;
		return buildNiceTimeTicksToDisplay(maxX, 5);
	}, [powerData]);

	const allSeries = useMemo(
		() => [
			...(showDiscrepencyCurve ? [discrepencyCurveSeries] : []),
			powerDataSeries,
			...(showDetectedSteps ? [detectedStepsSeries] : [])
		],
		[
			detectedStepsSeries,
			discrepencyCurveSeries,
			powerDataSeries,
			showDetectedSteps,
			showDiscrepencyCurve
		]
	);

	return (
		<>
			<ExpansionPanel>
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Box display="flex" flexDirection="column" {...stopClickFocusPropagation}>
						<Box
							display="flex"
							flexDirection="row"
							alignItems="center"
							minHeight="2em"
							width="100%"
							{...stopClickFocusPropagation}
						>
							<ActivityLoadHeader />
						</Box>
						<Box width="100%" {...stopClickFocusPropagation}>
							<SettingsFormGroup />
						</Box>
					</Box>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Box display="flex" flexDirection="column" width="100%">
						<FormGroup row>
							<FormControlLabel
								control={
									<Switch
										checked={showDiscrepencyCurve}
										onChange={e => setShowDiscrepencyCurve(e.target.checked)}
										name="showDCurve"
										color="secondary"
									/>
								}
								label="Show Discrepency Curve"
							/>
							<FormControlLabel
								control={
									<Switch
										checked={showDetectedSteps}
										onChange={e => setShowDetectedSteps(e.target.checked)}
										name="showSteps"
									/>
								}
								label="Show Steps"
							/>
						</FormGroup>
						<Box className="workout-creator-activity-data-chart">
							<XYPlot
								series={allSeries}
								xTickFormat={formatSecondsAsHHMMSS}
								xTickValues={activityXTicks}
								xAxisLabel="time"
								yAxisLabel="Power"
							/>
						</Box>
					</Box>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		</>
	);
};

export default ActivityLoader;
