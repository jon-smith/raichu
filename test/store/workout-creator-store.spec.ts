import { Action } from 'redux';
import { rootReducer } from 'state/reducers';
import * as Actions from 'state/workout-creator/workout-creator-actions';

const getInitialState = () => rootReducer(undefined, {} as Action);

describe('Workout Creator Store', () => {
	it('initial state should match snapshot', () => {
		const initialState = getInitialState();
		expect(initialState.workoutCreator).toMatchSnapshot();
	});

	it('updates intervals', () => {
		const store = getInitialState();

		const intervalsToSet = [{ color: '', length: 60, intensity: 1.2 }];
		const updatedStore = rootReducer(store, Actions.setIntervals(intervalsToSet));

		expect(updatedStore.workoutCreator.currentIntervals).toEqual(intervalsToSet);
		expect(updatedStore.workoutCreator.history.length).toEqual(2);
		expect(updatedStore.workoutCreator.currentHistoryPosition).toEqual(1);
	});

	it('doesnt change intervals if set to same', () => {
		const store = getInitialState();
		const intervalCopy = store.workoutCreator.currentIntervals.slice();

		const updatedStore = rootReducer(store, Actions.setIntervals(intervalCopy));

		// Confirm exact reference to original intervals
		expect(updatedStore.workoutCreator.currentIntervals).toBe(
			store.workoutCreator.currentIntervals
		);
		expect(updatedStore.workoutCreator.history.length).toEqual(1);
		expect(updatedStore.workoutCreator.currentHistoryPosition).toEqual(0);
	});

	it('doesnt change intervals if set to same with extra properties', () => {
		const store = getInitialState();
		const intervalsWithExtra = store.workoutCreator.currentIntervals.map(
			i => ({ ...i, extra: 100 } as Actions.Interval)
		);

		const updatedStore = rootReducer(store, Actions.setIntervals(intervalsWithExtra));

		// Confirm exact reference to original intervals
		expect(updatedStore.workoutCreator.currentIntervals).toBe(
			store.workoutCreator.currentIntervals
		);
		expect(updatedStore.workoutCreator.history.length).toEqual(1);
		expect(updatedStore.workoutCreator.currentHistoryPosition).toEqual(0);
	});

	it('undo redo', () => {
		const initialStore = getInitialState();

		const afterInitialUpdate = rootReducer(initialStore, Actions.setIntervals([]));
		expect(afterInitialUpdate.workoutCreator.currentIntervals).toEqual([]);

		const afterUndo = rootReducer(afterInitialUpdate, Actions.undo());
		expect(afterUndo.workoutCreator.currentIntervals).toEqual(
			initialStore.workoutCreator.currentIntervals
		);

		const afterRedo = rootReducer(afterUndo, Actions.redo());
		expect(afterRedo.workoutCreator.currentIntervals).toEqual(
			afterInitialUpdate.workoutCreator.currentIntervals
		);
	});
});
