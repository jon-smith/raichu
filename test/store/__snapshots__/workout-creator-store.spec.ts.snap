// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workout Creator Store initial state should match snapshot 1`] = `
Object {
  "currentHistoryPosition": 0,
  "currentIntervals": Array [
    Object {
      "color": "#eff3ff",
      "intensity": 0.3,
      "length": 60,
    },
    Object {
      "color": "#eff3ff",
      "intensity": 0.4,
      "length": 60,
    },
    Object {
      "color": "#eff3ff",
      "intensity": 0.5,
      "length": 60,
    },
    Object {
      "color": "#bdd7e7",
      "intensity": 0.6,
      "length": 300,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#3182bd",
      "intensity": 1.3,
      "length": 30,
    },
    Object {
      "color": "#6baed6",
      "intensity": 0.6,
      "length": 15,
    },
    Object {
      "color": "#bdd7e7",
      "intensity": 0.6,
      "length": 300,
    },
  ],
  "history": Array [
    Array [
      Object {
        "color": "#eff3ff",
        "intensity": 0.3,
        "length": 60,
      },
      Object {
        "color": "#eff3ff",
        "intensity": 0.4,
        "length": 60,
      },
      Object {
        "color": "#eff3ff",
        "intensity": 0.5,
        "length": 60,
      },
      Object {
        "color": "#bdd7e7",
        "intensity": 0.6,
        "length": 300,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#3182bd",
        "intensity": 1.3,
        "length": 30,
      },
      Object {
        "color": "#6baed6",
        "intensity": 0.6,
        "length": 15,
      },
      Object {
        "color": "#bdd7e7",
        "intensity": 0.6,
        "length": 300,
      },
    ],
  ],
  "selectedIndex": null,
}
`;
